File: 421.数组中两个数的最大异或值/421_example1.go
package LeetCode421

import "strings"

type trieNode struct {
	children []*trieNode
}

func constructNode() *trieNode {
	return &trieNode{children: make([]*trieNode, 2)}
}

// 将整数转为32位的二进制字符串，前面不为1用0补充
func toBinaryString(num int) string {
	if num == 0 {
		return "0"
	}
	bins := make([]rune, 32)
	mask := 1 << 31
	for i := 0; i < len(bins); i++ {
		if (mask & num) != 0 {
			bins[i] = '1'
		} else {
			bins[i] = '0'
		}
		mask >>= 1 // 向右移动一位，去掉一个最高位
	}
	return strings.TrimLeft(string(bins), "0") // 左边空白字符串用0补充
}

func findMaximumXOR(nums []int) int {
	maxNum := 0
	for _, n := range nums {
		if n > maxNum {
			maxNum = n
		}
	}
	l := len(toBinaryString(maxNum))
	mask := 1 << l
	bins := make([]string, len(nums))
	for i := 0; i < len(nums); i++ {
		bins[i] = toBinaryString(nums[i] | mask)[1:]
	}
	maxXor := 0
	root := constructNode()
	for _, bin := range bins {
		node, xorNode := root, root
		xor := 0
		for _, r := range bin {
			// 插入
			idx := r - '0'
			if node.children[idx] == nil {
				node.children[idx] = constructNode()
			}
			node = node.children[idx]

			// 计算xor
			if xorNode.children[1-idx] != nil {
				xor = (xor << 1) + 1
				xorNode = xorNode.children[1-idx]
			} else {
				xor = xor << 1
				xorNode = xorNode.children[idx]
			}
		}
		if maxXor < xor {
			maxXor = xor
		}
	}
	return maxXor
}

File: 421.数组中两个数的最大异或值/421_example1.py

# https: // leetcode-cn.com/problems/maximum-xor-of-two-numbers-in-an-array/solution/qian-zhui-shu-by-powcai/
# 异或运算其实就是 二进制下不进位的加法
# 思路一：贪心

# 看最高位是否异或为 1,

# 比如示例中：[3, 10, 5, 25, 2, 8] -> [00011,01010, 00101, 11001, 00010, 01000]

# 从左到右最高位参看，我们发现最高位0, 0, 0, 1, 0, 0，说明最大数至少为10000

# 还有一个公式 ，如果a ^ b = c，那么c ^ a = b

# 直接看代码，理解一下吧！
#


class Solution:
    def findMaximumXOR(self, nums: List[int]) -> int:
        res = 0
        mask = 0
        for i in range(32, -1, -1):
            mask = mask | (1 << i)
            # 记录前缀
            s = set()
            for num in nums:
                s.add(num & mask)  # 得到每一位与结果，能知道该位置是1还是0
            tmp = res | (1 << i)
            for t in s:
                if tmp ^ t in s:
                    res = tmp
                    break
        return res

# 前缀树


class Solution:
    def findMaximumXOR(self, nums: List[int]) -> int:
        if not nums:
            return 0
        # 创建前缀树
        root = {}
        for num in nums:
            cur = root
            for i in range(31, -1, -1):
                cur_bit = (num >> i) & 1
                cur.setdefault(cur_bit, {})
                cur = cur[cur_bit]

        res = float("-inf")
        # 按位找最大值
        for num in nums:
            cur = root
            cur_max = 0
            for i in range(31, -1, -1):
                cur_bit = (num >> i) & 1
                if cur_bit ^ 1 in cur:
                    cur_max += (1 << i)
                    cur = cur[cur_bit ^ 1]
                else:
                    cur = cur[cur_bit]
            res = max(res, cur_max)
        return res
